course-material/intro-to-sql/joining-data-in-sql.md

```diff
@@ -1,14 +1,397 @@
+
+There exists only two value types in the `accounts` and `district` table: "INT" and "VARCHAR". The "INT" value type indicates that the corresponding value is an integer, while "VARCHAR" represents a variable-length string that can contain various characters. These value types help SQL understand the appropriate operations that can be performed on each value. Alongside this, the second column of our tables also show if a variable is a primary key (PK) or foreign key (FK). We introduce these concepts below.
+
+#### What is a primary key and a foreign key?
+
+In a database, a <b>primary key</b> is a unique identifier for each record in a table. For instance, our `accounts` table has the primary key of "account_id". This makes sense because every single row in the `accounts` table corresponds to "account_id" which represents one single account. The `district` table has the primary key of "District ID". This means that each row under the `district` table represents one single district (or district id). So under the `accounts` table, there should not be any rows with the same "account_id" value. Similarly, the `district` table should not have any rows with the same "district_id" value. 
+
+A foreign key, on the other hand, establishes a relationship between two tables. It refers to the primary key of another table and helps connect the records across multiple tables. In our example, the "district_id" in the account table is a foreign key, indicating that it references the primary key of the district table. This allows us to associate each account with its corresponding district.
+
+By using primary and foreign keys, we establish relationships between tables, enabling us to perform joins and retrieve meaningful information by linking related data together. Let's jump straight into demonstrating these joins and how primary keys and foreign keys work.
+
+<!-- #region -->
+## Install - execute this once. 
+<b>Note:</b> If you are following these lessons locally and <b>not</b> on Google Colab, then there is no need to reinstall these packages.
+
+This code installs JupySQL, DuckDB, and Pandas in your environment. We will be using these moving forward.
+
+```{code-cell} ipython3
+%pip install jupysql --upgrade duckdb-engine pandas --quiet
+```
+
+## Load the data
+We extract the financial data by retrieving it from it's URL download link. The link may be a zip file (which it is in this case), so we extract the zip file and conver the .asc files to .csv files. Finally, we save converted data into a folder.
+
+```{code-cell} ipython3
+import csv
+import urllib.request
+import zipfile
+import os
+def extract_asc_to_csv(url, output_folder):
```

Comment:
Might be worth having a stand alone script .py that can be reused. Otherwise you need to make fixes to the functions in a lot of places